### PR TITLE
Add a command to clean up stale inflight records

### PR DESCRIPTION
**Pull request by HappyTetrahedron**
This adds the cleanup command, which finds old partner records in Odoo that still have the inflight flag set and deletes them.

This command will only work for the odoo8 storage backend, as it is only relevant there.

## Summary

...

## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: apiserver/billing/odoostorage/odoo/odoo8/odoo8.go [hunk 1]
```diff
@@ -274,6 +278,40 @@ func (s *oodo8Storage) Update(ctx context.Context, be *billingv1.BillingEntity)
...
+	if err != nil {
+		return err
+	}
+
+	ids := []int{}
+
+	for _, record := range inflightRecords {
+		updateTime := record.CreationTimestamp.ToTime()
+
+		if updateTime.Before(time.Now().Add(-1 * minAge)) {
```
Comment (bastjan) [inline apiserver/billing/odoostorage/odoo/odoo8/odoo8.go:303]:
```suggestion
		createTime := record.CreationTimestamp.ToTime()

		if createTime.Before(time.Now().Add(-1 * minAge)) {
```

File: cleanup.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	cmd := &cobra.Command{
+		Use: "cleanup",
+	}
+
+	odooUrl := cmd.Flags().String("billing-entity-odoo8-url", "http://localhost:8069", "URL of the Odoo instance to use for billing entities")
+	debugTransport := cmd.Flags().Bool("billing-entity-odoo8-debug-transport", false, "Enable debug logging for the Odoo transport")
+	countryList := cmd.Flags().String("billing-entity-odoo8-country-list", "countries.yaml", "Path to the country list file in the format of [{name: \"Germany\", code: \"DE\", id: 81},...]")
+	accountingContactDisplayName := cmd.Flags().String("billing-entity-odoo8-accounting-contact-display-name", "Accounting", "Display name of the accounting contact")
+	languagePreference := cmd.Flags().String("billing-entity-odoo8-language-preference", "en_US", "Language preference of the Odoo record")
+	paymentTerm := cmd.Flags().Int("billing-entity-odoo8-payment-term-id", 2, "Payment term ID of the Odoo record")
```
Comment (bastjan) [inline cleanup.go:26]: Don't think these are required

**Review by bastjan — COMMENTED** [review]
I'd move the cleanup function to it's own struct/class. There's almost no parameter overlap and it needs to be initiated twice anyways.

Putting everything in one struct makes it very hard to track which config parameters are required and makes extensions/refactoring harder.

File: apiserver/billing/odoostorage/odoo/odoo8/odoo8.go [hunk 3]
```diff
@@ -63,22 +65,36 @@ type Config struct {
...
+}
+
+type Odoo8Storage struct {
 	config Config
 
 	sessionCreator func(ctx context.Context) (client.QueryExecutor, error)
 }
 
-func (s *oodo8Storage) Get(ctx context.Context, name string) (*billingv1.BillingEntity, error) {
+type Odoo8StorageForCleanup struct {
```
[the remark this answers is not in the document]
Comment (bastjan) [reply apiserver/billing/odoostorage/odoo/odoo8/odoo8.go:93]: Maybe also its own file but that is personal preference

[the remark this answers is not in the document]
Comment (HappyTetrahedron) [reply apiserver/billing/odoostorage/odoo/odoo8/odoo8.go]: An excellent suggestion. 👍🏻

File: apiserver/billing/odoostorage/odoo/odoo8/odoo8.go [hunk 4]
```diff
@@ -274,6 +290,42 @@ func (s *oodo8Storage) Update(ctx context.Context, be *billingv1.BillingEntity)
 	return nil
 }
 
+func (s *FailedRecordScrubber) CleanupIncompleteRecords(ctx context.Context, minAge time.Duration) error {
+	// CleanupIncompleteRecords looks for partner records in Odoo that still have the "inflight" flag set despite being older than `minAge`. Those records are then deleted.
+	// Such records might come into existence due to a partially failed creation request.
```
Comment (bastjan) [inline apiserver/billing/odoostorage/odoo/odoo8/odoo8.go:295]:
```suggestion
// CleanupIncompleteRecords looks for partner records in Odoo that still have the "inflight" flag set despite being older than `minAge`. Those records are then deleted.
// Such records might come into existence due to a partially failed creation request.
func (s *FailedRecordScrubber) CleanupIncompleteRecords(ctx context.Context, minAge time.Duration) error {
```

Go != python 😉😅

**Pull request merged**